src/simulators/statevector/qubitvector_thrust.hpp

```diff
@@ -0,0 +1,3146 @@
+	}
+	else{
+		int iDev;
+
+#pragma omp parallel for
+		for(iDev=0;iDev<m_nDevParallel;iDev++){
+			uint_t is,ie;
+			is = n * iDev / m_nDevParallel;
+			ie = n * (iDev+1) / m_nDevParallel;
+
+			auto ci = thrust::counting_iterator<uint_t>(0);
+
+			cudaSetDevice(iDev);
+			thrust::for_each(thrust::device, ci + is, ci + ie, func);
+		}
+	}
+}
+
+template <typename data_t>
+template <typename UnaryFunction>
+double QubitVectorThrust<data_t>::apply_sum_lambda(UnaryFunction func,uint_t n) const
+{
+	double ret = 0.0;
+
+	if(m_nDevParallel == 1){
+		auto ci = thrust::counting_iterator<uint_t>(0);
+
+		ret = thrust::transform_reduce(thrust::device, ci, ci+n, func,0.0,thrust::plus<double>());
+	}
+	else{
+		int iDev;
+
+#pragma omp parallel for reduction(+:ret)
+		for(iDev=0;iDev<m_nDevParallel;iDev++){
+			uint_t is,ie;
+			is = n * iDev / m_nDevParallel;
+			ie = n * (iDev+1) / m_nDevParallel;
+
+			auto ci = thrust::counting_iterator<uint_t>(0);
+
+			cudaSetDevice(iDev);
+			ret += thrust::transform_reduce(thrust::device, ci + is, ci + ie, func,0.0,thrust::plus<double>());
+		}
+	}
+	return ret;
+}
+
+template <typename data_t>
+void QubitVectorThrust<data_t>::apply_matrix(const reg_t &qubits,
+                                       const cvector_t<double> &mat)
+{
+	const size_t N = qubits.size();
+	uint_t size;
+
+//	printf(" Mat Mult : %d\n",N);
+
+#ifdef QASM_TIMING
+	TimeStart(QS_GATE_MULT);
+#endif
+	if(N == 1){
+		size = data_size_ >> 1;
+		apply_lambda(matMult2x2_lambda<data_t>((thrust::complex<data_t>*)&data_[0],(thrust::complex<double>*)&mat[0],qubits[0]), size);
+	}
+	else if(N == 2){
+		size = data_size_ >> 2;
+		apply_lambda(matMult4x4_lambda<data_t>((thrust::complex<data_t>*)&data_[0],(thrust::complex<double>*)&mat[0],qubits[0],qubits[1]), size);
+	}
+	else{
+		thrust::complex<double>* pMat;
+
+		uint_t i,matSize;
+		matSize = 1ull << N;
+
+		allocate_buffers(N);
+
+		pMat = m_pMatDev;
+
+#pragma omp parallel for 
+		for(i=0;i<matSize*matSize;i++){
+			m_pMatDev[i] = mat[i];
+		}
+
+		if(N == 3){
+			size = data_size_ >> 3;
+			apply_lambda(matMult8x8_lambda<data_t>((thrust::complex<data_t>*)&data_[0],pMat,qubits[0],qubits[1],qubits[2]), size);
+		}
+		else if(N == 4){
+			size = data_size_ >> 4;
+			apply_lambda(matMult16x16_lambda<data_t>((thrust::complex<data_t>*)&data_[0],pMat,qubits[0],qubits[1],qubits[2],qubits[3]), size);
+		}
+		else{
+			size = data_size_ >> N;
+			apply_lambda(matMultNxN_LU_lambda<data_t>((thrust::complex<data_t>*)&data_[0],pMat,m_pUintBuf,(uint_t*)&qubits[0],N), size);
+		}
+	}
+
+#ifdef QASM_TIMING
+	TimeEnd(QS_GATE_MULT);
+#endif
+
+}
+
+template <typename data_t>
+void QubitVectorThrust<data_t>::apply_multiplexer(const reg_t &control_qubits,
+                                            const reg_t &target_qubits,
+                                            const cvector_t<double>  &mat) {
+	printf(" apply_multiplexer NOT SUPPORTED : %d, %d \n",control_qubits.size(),target_qubits.size());
```

Comment:
Can you throw an exception here instead of printing a warning. This will let the simulator catch and return the error message to qiskit users in result object.

Reply:
I added message by throwing an exception